### ISSUE TITLE
[1.12.2] Crash from Trying to Bucket Fluid Ethanol

### ISSUE THREAD

**Issue by Vyraal1** [turbodiesel4598/NuclearCraft#253]
https://hastebin.com/ibekiwuzex.sql

Used JEI in creative testing to cheat myself a fluid source block (it didn't give me the bucket variant of the fluid for some reason), I tried placing it in world and then tried to bucket it so I could figure out what it's the forge thought the liquid was (eg liquid:ethanol) and ended up crashing.

**Comment by turbodiesel4598** [comment]
I am unable to recreate this bug - it may have something to do with the large modpack you're using or 2.9d being built on an old Forge version. A bug-fixing update's coming soon so we will be able to see if it is the latter.

**Comment by HenryLoenwind** [comment]
From further testing it looks like you create an Ethanol `Fluid` and use it, but do not register it with the FluidRegistry for some reason. As there *is* some Ethanol in the Registry, your Fluid partially works, until it crashes.

**ALWAYS** register your `Fluid`s. The registry's job is to figure out which Fluid implementation will be used.

**Comment by turbodiesel4598** [comment]
So something like
`FluidRegistry.addBucketForFluid(fluid);
if (!FluidRegistry.isFluidRegistered(fluid)) FluidRegistry.registerFluid(fluid);`
would be fine?

**Comment by HenryLoenwind** [comment]
no.

isFluidRegistered() only checks if there is any fluid with that name registered. What you need is

```
if (!FluidRegistry.registerFluid(fluid)) {
   fluid = FluidRegistry.getFluid(fluid.getName());
} # put an 'else' here if you don't want to force other mod's fluid to have a bucket
if (!FluidRegistry.getBucketFluids().contains(fluid)) {
    FluidRegistry.addBucketForFluid(fluid);
 }
```

**Comment by turbodiesel4598** [comment]
Riiiight, I see now - that clears a lot of things up.

**Comment by HenryLoenwind** [comment]
PS: I added an explanation above

**Comment by turbodiesel4598** [comment]
[Fixed in version 2.10c :)](https://minecraft.curseforge.com/projects/nuclearcraft-mod/files/2548550)
